synapse/handlers/room.py

```diff
@@ -150,6 +207,15 @@ def upgrade_room(self, requester, old_room_id, new_version):
                 requester, old_room_id, new_room_id, old_room_state,
             )
 
+            # Remove the pending request and the entry for this room id if
+            # necessary
+            self.currently_upgrading_rooms[old_room_id]["pending_requests"] -= 1
+            if self.currently_upgrading_rooms[old_room_id]["pending_requests"] <= 0:
+                del self.currently_upgrading_rooms[old_room_id]
+            else:
+                # Save this response if another request is waiting on it
+                self.upgrade_responses[old_room_id] = new_room_id
```

Comment:
This should be in a `try... finally...`  to properly handle the case of exceptions being thrown.